File: angel-ps/psf/src/main/java/com/tencent/angel/ml/matrix/psf/aggr/Equal.java
/*
 * Tencent is pleased to support the open source community by making Angel available.
 *
 * Copyright (C) 2017 THL A29 Limited, a Tencent company. All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 *
 * https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

package com.tencent.angel.ml.matrix.psf.aggr;

import com.tencent.angel.ml.matrix.psf.aggr.enhance.BinaryAggrFunc;
import com.tencent.angel.ml.matrix.psf.aggr.enhance.ScalarAggrResult;
import com.tencent.angel.ml.matrix.psf.aggr.enhance.ScalarPartitionAggrResult;
import com.tencent.angel.ml.matrix.psf.get.base.GetResult;
import com.tencent.angel.ml.matrix.psf.get.base.PartitionGetResult;
import com.tencent.angel.ps.impl.matrix.ServerDenseDoubleRow;
import com.tencent.angel.ps.impl.matrix.ServerSparseDoubleLongKeyRow;
import it.unimi.dsi.fastutil.longs.Long2DoubleOpenHashMap;
import it.unimi.dsi.fastutil.longs.LongOpenHashSet;
import it.unimi.dsi.fastutil.longs.LongSemiIndirectHeaps;
import it.unimi.dsi.fastutil.longs.LongSet;

import java.nio.DoubleBuffer;
import java.util.List;

/**
 * `Equal` judges if two rows is equal.
 */
public final class Equal extends BinaryAggrFunc {

  public Equal(int matrixId, int rowId1, int rowId2) {
    super(matrixId, rowId1, rowId2);
  }

  public Equal() {
    super();
  }

  @Override
  protected double doProcessRow(ServerDenseDoubleRow row1, ServerDenseDoubleRow row2) {
    double isEqual = 1.0;
    DoubleBuffer data1 = row1.getData();
    DoubleBuffer data2 = row2.getData();
    int size = row1.size();
    for (int i = 0; i < size; i++) {
      if (Math.abs(data1.get(i) - data2.get(i)) > 1e-10) {
        isEqual = 0.0;
        break;
      }
    }
    return isEqual;
  }

  @Override
  protected double doProcessRow(ServerSparseDoubleLongKeyRow row1, ServerSparseDoubleLongKeyRow row2) {
    Long2DoubleOpenHashMap data1 = row1.getIndex2ValueMap();
    Long2DoubleOpenHashMap data2 = row2.getIndex2ValueMap();

    if (data1.defaultReturnValue() != data2.defaultReturnValue()) {
      return 0.0;
    }
    LongSet keys = new LongOpenHashSet(data1.keySet());
    keys.addAll(data2.keySet());

    if (keys.size() != data1.keySet().size() || keys.size() != data2.keySet().size()) {
      return 0.0;
    }

    for (long key: keys) {
      if (Math.abs(data1.get(key) - data2.get(key)) > 1e-11) {
        return 0.0;
      }
    }
    return 1.0;
  }

  @Override
  public GetResult merge(List<PartitionGetResult> partResults) {
    double isEqual = 1.0;
    for (PartitionGetResult partResult : partResults) {
      if (partResult != null) {
        double res = ((ScalarPartitionAggrResult) partResult).result;
        if (res != 1.0) {
          isEqual = res;
          break;
        }
      }
    }

    return new ScalarAggrResult(isEqual);
  }

}

File: angel-ps/psf/src/main/java/com/tencent/angel/ml/matrix/psf/aggr/Equal.scala
/*
 * Tencent is pleased to support the open source community by making Angel available.
 *
 * Copyright (C) 2017-2018 THL A29 Limited, a Tencent company. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in 
 * compliance with the License. You may obtain a copy of the License at
 *
 * https://opensource.org/licenses/Apache-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 *
 */


package com.tencent.angel.ml.matrix.psf.aggr

;

import scala.collection.JavaConversions._
import com.tencent.angel.exception.AngelException
import com.tencent.angel.ml.math2.storage.{IntDoubleDenseVectorStorage, IntFloatDenseVectorStorage, IntIntDenseVectorStorage, IntLongDenseVectorStorage}
import com.tencent.angel.ml.matrix.psf.aggr.enhance.{BinaryAggrFunc, ScalarAggrResult, ScalarPartitionAggrResult}
import com.tencent.angel.ml.matrix.psf.get.base.{GetResult, PartitionGetResult}
import com.tencent.angel.ps.storage.vector._

/**
  * `Equal` judges if two rows is equal.
  */
class Equal(matrixId: Int, rowId1: Int, rowId2: Int) extends BinaryAggrFunc(matrixId: Int, rowId1: Int, rowId2: Int) {
  val equal = 1.0
  val notEqual = 0.0
  val equalLimit = 1e-10

  def this() = this(-1, -1, -1)

  override def merge(partResults: java.util.List[PartitionGetResult]): GetResult = {
    var sum = 1.0
    for (partResult <- partResults) {
      if (partResult != null) {
        sum = math.min(sum, partResult.asInstanceOf[ScalarPartitionAggrResult].result)
      }
    }
    new ScalarAggrResult(sum)
  }

  override protected def processRows(row1: ServerRow, row2: ServerRow): Double = {
    val eq = if (row1.isDense && row2.isDense) {
      row1 match {
        case r1: ServerIntDoubleRow => ServerRowUtils.getVector(r1).getStorage.asInstanceOf[IntDoubleDenseVectorStorage].getValues.zip(ServerRowUtils.getVector(row2).getStorage.asInstanceOf[IntDoubleDenseVectorStorage].getValues)
          .forall { case (a, b) => math.abs(a - b) < equalLimit }
        case r1: ServerIntFloatRow => ServerRowUtils.getVector(r1).getStorage.asInstanceOf[IntFloatDenseVectorStorage].getValues.zip(ServerRowUtils.getVector(row2).getStorage.asInstanceOf[IntFloatDenseVectorStorage].getValues)
          .forall { case (a, b) => math.abs(a - b) < equalLimit }
        case r1: ServerIntLongRow => ServerRowUtils.getVector(r1).getStorage.asInstanceOf[IntLongDenseVectorStorage].getValues.zip(ServerRowUtils.getVector(row2).getStorage.asInstanceOf[IntLongDenseVectorStorage].getValues)
          .forall { case (a, b) => math.abs(a - b) < equalLimit }
        case r1: ServerIntIntRow => ServerRowUtils.getVector(r1).getStorage.asInstanceOf[IntIntDenseVectorStorage].getValues.zip(ServerRowUtils.getVector(row2).getStorage.asInstanceOf[IntIntDenseVectorStorage].getValues)
          .forall { case (a, b) => math.abs(a - b) < equalLimit }
        case _ => throw new AngelException("should not come here!")
      }
    } else {
      ServerRowUtils.getVector(row1).sub(ServerRowUtils.getVector(row2)).ifilter(equalLimit).sum() == 0
    }
    if (eq) equal else notEqual
  }
}
